src/main/kotlin/org/rust/ide/navigation/goto/RsGotoSuperHandler.kt

```diff
@@ -32,22 +34,38 @@ class RsGotoSuperHandler : LanguageCodeInsightActionHandler {
 
 // public for testing
 fun gotoSuperTarget(source: PsiElement): NavigatablePsiElement? {
-    val modOrAbstractable = PsiTreeUtil.getNonStrictParentOfType(
+    val it = PsiTreeUtil.getNonStrictParentOfType(
         source,
         RsAbstractable::class.java,
-        RsMod::class.java
-    ) ?: return null
+        RsMod::class.java,
+        TomlFile::class.java
+    )
+    if (it is RsAbstractable) return when {
+        it.owner.isTraitImpl -> it.superItem
+        else -> gotoSuperTarget(it.parent)
+    }
 
-    if (modOrAbstractable is RsAbstractable) {
-        return if (modOrAbstractable.owner.isTraitImpl) {
-            modOrAbstractable.superItem
+    if (it is RsMod) {
+        return if (it is RsFile) {
+            if (it.isCrateRoot) {
+                it.containingCargoPackage
+                    ?.rootDirectory
+                    ?.resolve("Cargo.toml")
+                    ?.toCargoTomlPsiFile(source.project)
+            } else {
+                it.declaration
+            }
         } else {
-            gotoSuperTarget(modOrAbstractable.parent)
+            it.`super`
         }
     }
 
-    return when (val mod = modOrAbstractable as RsMod) {
-        is RsFile -> mod.declaration
-        else -> mod.`super`
-    }
+    return null
+}
+
+fun Path.toCargoTomlPsiFile(project: Project): PsiFile? {
+    val virtualFile = LocalFileSystem.getInstance().findFileByPath("$this") ?: return null
+    val psiFile = PsiManager.getInstance(project).findFile(virtualFile) ?: return null
+    if (psiFile.fileType != TomlFileType) return null
```

Comment:
Looks like this will crash the plugin if no `toml` plugin installed

Reply:
Oops, missed that one. Fixed now